File: seminar-5/queue/cpp/queue_by_list.cpp
#include <stdexcept>

template<typename T>
class Queue {
private:
    T* data;
    int front;
    int size_;
    int capacity;

    void resize(int newCapacity) {
        T* newData = new T[newCapacity];
        for (int i = 0; i < size_; i++) {
            newData[i] = data[front + i];
        }
        delete[] data;
        data = newData;
        front = 0;
        capacity = newCapacity;
    }

public:
    Queue(int initialCapacity = 4)
        : front(0), size_(0), capacity(initialCapacity) {
        data = new T[capacity];
    }

    ~Queue() {
        delete[] data;
    }

    void enqueue(const T& value) {
        if (front + size_ == capacity) {
            resize(capacity * 2);
        }
        data[front + size_] = value;
        size_++;
    }

    T dequeue() {
        if (isEmpty()) {
            throw std::runtime_error("dequeue from empty queue");
        }

        T value = data[front];
        front++;
        size_--;

        if (size_ == 0) {
            front = 0;
        }

        return value;
    }

    T peek() const {
        if (isEmpty()) {
            throw std::runtime_error("peek from empty queue");
        }
        return data[front];
    }

    bool isEmpty() const {
        return size_ == 0;
    }

    int size() const {
        return size_;
    }
};

File: seminar-5/queue/python/queue_by_list.py
class Queue:
    """Очередь (FIFO) на динамическом массиве с ручным расширением."""

    def __init__(self, capacity=4):
        """
        Создаёт пустую очередь.

        :param capacity: начальная ёмкость массива
        """
        self._data = [None] * capacity
        self._front = 0
        self._size = 0

    def _resize(self, new_capacity):
        """
        Изменяет ёмкость массива.
        """
        new_data = [None] * new_capacity
        for i in range(self._size):
            new_data[i] = self._data[self._front + i]
        self._data = new_data
        self._front = 0

    def enqueue(self, value):
        """
        Добавляет элемент в конец очереди.
        """
        # если нет места в массиве — расширяем
        if self._front + self._size == len(self._data):
            self._resize(len(self._data) * 2)

        index = self._front + self._size
        self._data[index] = value
        self._size += 1

    def dequeue(self):
        """
        Удаляет и возвращает элемент из начала очереди.
        """
        if self.is_empty():
            raise IndexError("dequeue from empty queue")

        value = self._data[self._front]
        self._data[self._front] = None
        self._front += 1
        self._size -= 1

        # если очередь опустела — сбрасываем front
        if self._size == 0:
            self._front = 0

        return value

    def peek(self):
        """
        Возвращает первый элемент без удаления.
        """
        if self.is_empty():
            raise IndexError("peek from empty queue")
        return self._data[self._front]

    def is_empty(self):
        return self._size == 0

    def size(self):
        return self._size

    def __len__(self):
        return self._size

    def __bool__(self):
        return self._size != 0

    def __repr__(self):
        front = None if self.is_empty() else self._data[self._front]
        return f"Queue(size={self._size}, front={front!r})"

    def __iter__(self):
        for i in range(self._size):
            yield self._data[self._front + i]

    def to_list(self):
        return list(self)
